cmd/docker-notation/push.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/distribution/distribution/v3/manifest/schema2"
+	"github.com/notaryproject/notation/cmd/docker-notation/docker"
+	"github.com/notaryproject/notation/pkg/config"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/urfave/cli/v2"
+)
+
+var pushCommand = &cli.Command{
+	Name:      "push",
+	Usage:     "Push an image or a repository to a registry",
+	ArgsUsage: "[<reference>]",
+	Action:    pushImage,
+}
+
+func pushImage(ctx *cli.Context) error {
+	if err := passThroughIfNotationDisabled(ctx); err != nil {
+		return err
+	}
+
+	desc, err := pushImageAndGetOCIDescriptor(ctx)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Pushing signature")
+	sigDigests, err := config.SignatureDigests(desc.Digest)
+	if err != nil {
+		return err
+	}
+	if len(sigDigests) == 0 {
+		return errors.New("no signatures found")
+	}
+
+	client, err := docker.GetSignatureRepository(ctx.Context, ctx.Args().First())
+	if err != nil {
+		return err
+	}
+	pushSignature := func(sigDigest digest.Digest) error {
+		sigPath := config.SignaturePath(desc.Digest, sigDigest)
+		sig, err := os.ReadFile(sigPath)
+		if err != nil {
+			return err
+		}
+
+		sigDesc, err := client.Put(ctx.Context, sig)
+		if err != nil {
+			return err
+		}
+
+		artifactDesc, err := client.Link(ctx.Context, desc, sigDesc)
```

Comment:
What doe `Link` do here?

Reply:
It pushes an artifact manifest, which contains `sigDesc` and references `desc`.